arclightning_backend/src/main.rs

```diff
@@ -0,0 +1,201 @@
+extern crate futures;
+extern crate hyper;
+#[macro_use]
+extern crate serde_derive;
+#[macro_use]
+extern crate serde_json;
+extern crate toml;
+
+use futures::future;
+
+use hyper::rt::{Future, Stream};
+use hyper::service::service_fn;
+use hyper::{Body, Method, Request, Response, Server, StatusCode};
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+use std::path::PathBuf;
+use std::sync::{Arc, Mutex};
+
+// after this concept is further understood, will switch to 'Either'
+type ResponseFuture = Box<Future<Item = Response<Body>, Error = hyper::Error> + Send>;
+
+// using PartialEq for unit tests
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+// Using clone in a unit test atm.  Might not be necessary
+struct Game {
+    name: String,
+    description: String,
+    genres: Vec<String>,
+    thumbnail_path: PathBuf,
+    exe_path: PathBuf,
+}
+
+fn router(request: Request<Body>) -> ResponseFuture {
+    let mut response = Response::new(Body::empty());
+
+    match (request.method(), request.uri().path()) {
+        (&Method::GET, "/api/v1/list_games") => {
+            *response.body_mut() = Body::from("test");
+        }
+        _ => {
+            *response.status_mut() = StatusCode::NOT_FOUND;
+        }
+    }
+    Box::new(future::ok(response))
+}
+
+fn toml_to_hashmap(toml_filepath: PathBuf) -> HashMap<String, Game> {
+    let mut games_toml = String::new();
+    let mut file = File::open(&toml_filepath).expect("Could not find .toml file");
+    file.read_to_string(&mut games_toml)
+        .unwrap_or_else(|err| panic!("Error while reading .toml file: [{}]", err));
+    toml::from_str(&games_toml).unwrap()
+}
+
+fn main() {
+    // Read initial games toml config
+    let toml_filepath: PathBuf = ["test_files", "test_games.toml"].iter().collect();
+
+    // Store games locally on server
+    let games: HashMap<String, Game> = toml_to_hashmap(toml_filepath);
+    let games_data = Arc::new(Mutex::new(games.clone()));
+
+    // Host server
+    let addr = ([127, 0, 0, 1], 3000).into();
+
+    let server = Server::bind(&addr)
+        .serve(|| service_fn(router))
+        .map_err(|err| eprintln!("server error: {}", err));
+
+    println!("Listening on http://{}", addr);
+    hyper::rt::run(server);
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    // TODO
+    // not implemented yet, test sending the games as a json as payload
+    // Don't know how to assert here, if feasible
+    fn test_games_endpoint() {
+        use serde_json::value::Value;
+        use std::fs::File;
+        use std::io::Read;
+        use std::sync::{Arc, Mutex};
```

Comment:
There seem to be a lot of `use` statements that are duplicates of the ones at the top of the file here.